format/bplist/bplist.go

```diff
@@ -104,82 +104,101 @@ func decodeItem(d *decode.D, p *plist) {
+		return d.FieldF("value", n*8, scalar.DescriptionTimeFn(scalar.S.TryActualF, cocoaTimeEpochDate, time.RFC3339))
 	case elementTypeData:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldRawLen("value", int64(n*8))
+		return d.FieldRawLen("value", int64(n*8))
 	case elementTypeASCIIString:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldUTF8("value", int(n))
+		return d.FieldUTF8("value", int(n))
 	case elementTypeUnicodeString:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldUTF16("value", int(n))
+		return d.FieldUTF16("value", int(n))
 	case elementTypeUID:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldUBigInt("value", int(n)).Uint64()
+		return d.FieldUBigInt("value", int(n)).Uint64()
 	case elementTypeArray:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldStructNArray("entries", "entry", int64(n),
+		return d.FieldStructNArray("entries", "entry", int64(n),
 			func(d *decode.D) {
-				idx := d.FieldU8("object_index")
-				decodeReference(d, p, idx)
+				idx := d.FieldU("object_index", int(p.t.objRefSize)*8)
+				p.decodeReference(d, idx)
 			})
 	case elementTypeSet:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldStructNArray("entries", "entry", int64(n),
+		return d.FieldStructNArray("entries", "entry", int64(n),
 			func(d *decode.D) {
-				idx := d.FieldU8("object_index")
-				decodeReference(d, p, idx)
+				idx := d.FieldU("object_index", int(p.t.objRefSize)*8)
+				p.decodeReference(d, idx)
 			})
 	case elementTypeDict:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldStructNArray("entries", "entry", int64(n),
+		return d.FieldStructNArray("entries", "entry", int64(n),
 			func(d *decode.D) {
 				var ki, vi uint64
-				ki = d.FieldU8("key_index")
+				ki = d.FieldU("key_index", int(p.t.objRefSize)*8)
 				d.SeekRel(int64((n-1)*p.t.objRefSize)*8, func(d *decode.D) {
-					vi = d.FieldU8("value_index")
+					vi = d.FieldU("value_index", int(p.t.objRefSize)*8)
 				})
 				d.FieldStruct("key", func(d *decode.D) {
-					decodeReference(d, p, ki)
+					k := p.decodeReference(d, ki)
+					if _, ok := k.(string); !ok {
+						d.Errorf("non-string key in dictionary")
+					}
 				})
 				d.FieldStruct("value", func(d *decode.D) {
-					decodeReference(d, p, vi)
+					p.decodeReference(d, vi)
 				})
 			})
 	default:
 		d.Errorf("unknown type marker: %d", m)
+		return nil
 	}
 }
 
-func decodeReference(d *decode.D, p *plist, idx uint64) {
-	if idx > uint64(len(p.o)) {
+func (pl *plist) decodeReference(d *decode.D, idx uint64) any {
+	if idx > uint64(len(pl.o)) {
 		// prevent a panic
-		d.Errorf("index %d out of bounds for object table size %d", idx, len(p.o))
-		return
+		d.Errorf("index %d out of bounds for object table size %d", idx, len(pl.o))
+		return nil
 	}
 
-	d.SeekAbs(int64(p.o[idx]*8), func(d *decode.D) {
-		decodeItem(d, p)
+	if pl.indexIsInStack(idx) {
+		d.Errorf("recursion detected: object %d already decoded in stack %v", idx, pl.objectStack)
```

Comment:
This should probably be `d.Fatalf`, difference is that `d.Errorf` is ignore when using `-o force=true`. It's a way to skip magic checks and other things that wont lead to non-recoverable errors or loop etc

Reply:
Maybe Errorf should be renamed to make it more clear? Assert? i think error/fatal comes from how go test framework works

Reply:
I think `Errorf` is okay for now, feels natural to keep it in line with `go test` syntax